src/awkward/_v2/cpp-headers/GrowableBuffer.h

```diff
@@ -0,0 +1,298 @@
+      ptr_.clear();
+      ptr_.push_back(std::unique_ptr<PRIMITIVE>(new PRIMITIVE[initial_]));
+    }
+
+    /// @brief Inserts one `datum` into the panel, possibly triggering
+    /// allocation of a new panel.
+    ///
+    /// This increases the #length by 1; if the new #length is larger than
+    /// #reserved, a new panel will be allocated.
+    void
+    append(PRIMITIVE datum) {
+      if (length_[ptr_.size()-1] == reserved_[ptr_.size()-1]) {
+        add_panel(reserved_[ptr_.size()-1]);
+      }
+      fill_panel(datum);
+    }
+
+    /// @brief Inserts an entire array into the panel(s), possibly triggering
+    /// allocation of a new panel.
+    ///
+    /// If the size is larger than the empty slots in the current panel, then,
+    /// first, the empty slots are filled and then a new panel will be allocated
+    /// for the rest of the array elements.
+    void
+    extend(PRIMITIVE* ptr, size_t size) {
+      size_t empty_slots = reserved_[ptr_.size()-1] - length_[ptr_.size()-1];
+      if (size > empty_slots) {
+        for (size_t i = 0; i < empty_slots; i++) {
+          fill_panel(ptr[i]);
+        }
+        add_panel(size - empty_slots > reserved_[ptr_.size()-1] ?
+                  size - empty_slots : reserved_[ptr_.size()-1]);
+        for (size_t i = empty_slots; i < size; i++) {
+          fill_panel(ptr[i]);
+        }
+      }
+      else {
+        for (size_t i = 0; i < size; i++) {
+          fill_panel(ptr[i]);
+        }
+      }
+    }
+
+    /// @brief Like append, but the type signature returns the reference to `PRIMITIVE`.
+    PRIMITIVE& append_and_get_ref(PRIMITIVE datum) {
+      append(datum);
+      return (&*ptr_.back())[length_.back()];
+    }
+
+    /// @brief Copies and concatenates all accumulated data from multiple panels to one
+    /// contiguously allocated `external_pointer`.
+    void
+    concatenate(PRIMITIVE* external_pointer) const noexcept {
+      size_t next_panel = 0;
+      for (size_t i = 0;  i < ptr_.size();  i++) {
+        memcpy(external_pointer + next_panel, reinterpret_cast<void*>(ptr_[i].get()), length_[i]*sizeof(PRIMITIVE));
+        next_panel += length_[i];
+      }
+    }
+
+    /// @brief Checks whether the array is contiguous.
+    size_t is_contiguous() {
+      return (ptr_.size() == 1);
+    }
+
+    /// @brief Takes one (possibly multi-panels) GrowableBuffer<FROM_PRIMITIVE>
+    /// and makes another (one panel) GrowableBuffer<TO_PRIMITIVE>.
+    ///
+    /// Used to change the data type of buffer content from `FROM_PRIMITIVE`
+    /// to `TO_PRIMITIVE` for building arrays.
+    template<typename FROM_PRIMITIVE>
+    static GrowableBuffer<PRIMITIVE>
+    fill_from(GrowableBuffer<FROM_PRIMITIVE>& from_buffer) {
+      auto length = from_buffer.length();
+      auto ptr = std::unique_ptr<PRIMITIVE>(new PRIMITIVE[length]);
+      PRIMITIVE* rawptr = ptr.get();
+      int64_t k = 0;
+      for (size_t i = 0;  i < from_buffer.ptr_.size(); i++) {
```

Comment:
I think, you need to add accessors to private members: ptr_ and length_

Reply:
Do we need to expose the length_ of every panel? As currently, we already have a method that returns accumulated length.

Reply:
Isn't the fact that GrowableBuffer is implemented by panels an internal detail that you don't want other code depending on? (For instance, if we find in the future that some mix of panels and reallocation is ideal... not that I think that will happen.) Exposing either `ptr_` or `length_` would reveal the panels.

Reply:
Yes, I agree. Exposing `ptr_` and `length_` should be avoided.